### PR TITLE
IGNITE-21013 Support encryption of cache dumps

### PR DESCRIPTION
**Pull request by nizhikov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

**Review by timoninmaxim — COMMENTED** [review]
Let's clarify in docs difference between snapshots and dumps when we talk about encryption. Snapshots can contain encrypted pages, while Dumps encrypt entry by self.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteSnapshotManager.java [hunk 1]
```diff
@@ -1229,9 +1234,10 @@ private IgniteInternalFuture<SnapshotOperationResponse> initLocalFullSnapshot(
                     blts,
                     res.parts(),
                     res.snapshotPointer(),
-                    cctx.gridConfig().getEncryptionSpi().masterKeyDigest(),
+                    encSpi.masterKeyDigest(),
```
[the remark this answers is not in the document]
Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteSnapshotManager.java:1237]: This not changed in the scope of the PR. So I propose to keep it as is.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotMetadata.java [hunk 2]
```diff
@@ -139,7 +142,8 @@ public SnapshotMetadata(
         @Nullable WALPointer snpRecPtr,
         @Nullable byte[] masterKeyDigest,
         boolean onlyPrimary,
-        boolean dump
+        boolean dump,
+        byte[] encKey
```
[the remark this answers is not in the document]
Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotMetadata.java:146]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/CreateDumpFutureTask.java [hunk 3]
```diff
@@ -122,6 +126,12 @@ public class CreateDumpFutureTask extends AbstractCreateSnapshotFutureTask imple
      */
     private final ConcurrentMap<Long, ByteBuffer> thLocBufs = new ConcurrentHashMap<>();
 
+    /**
+     * Map shared across all instances of {@link PartitionDumpContext} and {@link DumpEntrySerializer}.
+     * Used to store encrypted data in case dump encrypted.
+     */
+    private final ConcurrentMap<Long, ByteBuffer> encThLocBufs;
```
[the remark this answers is not in the document]
Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/CreateDumpFutureTask.java:133]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/Dump.java [hunk 4]
```diff
@@ -261,7 +280,22 @@ public DumpedPartitionIterator iterator(String node, int group, int part) {
...
 
-        DumpEntrySerializer serializer = new DumpEntrySerializer(thLocBufs);
+        SnapshotMetadata meta = metadata.stream().filter(m -> Objects.equals(m.folderName(), node)).findFirst().orElseGet(null);
+
+        boolean encrypted = meta.encryptionKey() != null;
+
+        if (encrypted && !Arrays.equals(meta.masterKeyDigest(), encSpi.masterKeyDigest())) {
+            throw new IllegalArgumentException("Dump '" +
+                meta.snapshotName() + "' has different master key digest. To restore this " +
+                "snapshot, provide the same master key.");
```
[the remark this answers is not in the document]
Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/Dump.java:290]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/DumpEntrySerializer.java [hunk 5]
```diff
@@ -184,38 +257,77 @@ public ByteBuffer writeToBuffer(
...
+
+        if (buf == null)
+            return null;
+
+        if (encSpi != null) {
+            ByteBuffer plainBuf = threadLocalBuffer(encThLocBufs);
+
+            plainBuf.limit(plainBuf.capacity());
+
+            if (plainBuf.capacity() < buf.limit()) {
```
[the remark this answers is not in the document]
Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/DumpEntrySerializer.java:270]: Fixed

**Review by timoninmaxim — APPROVED** [review]

**Pull request merged**